File: go/test/member_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/trello-sdk/go"
	"github.com/voxgig-sdk/trello-sdk/go/core"

	vs "github.com/voxgig-sdk/trello-sdk/go/utility/struct"
)

func TestMemberEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Member(nil)
		if ent == nil {
			t.Fatal("expected non-nil MemberEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"member": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Member(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Member(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := memberBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "member." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TRELLO_TEST_MEMBER_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		memberRef01Ent := client.Member(nil)
		memberRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "member"}), "member_ref01"))
		memberRef01Data["action_id"] = setup.idmap["action01"]
		memberRef01Data["board_id"] = setup.idmap["board01"]
		memberRef01Data["enterpris_id"] = setup.idmap["enterpris01"]
		memberRef01Data["organization_id"] = setup.idmap["organization01"]
		memberRef01Data["token_id"] = setup.idmap["token01"]

		memberRef01DataResult, err := memberRef01Ent.Create(memberRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		memberRef01Data = core.ToMapAny(entityData(memberRef01DataResult))
		if memberRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if memberRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		memberRef01Match := map[string]any{
			"organization_id": setup.idmap["organization01"],
		}

		memberRef01ListResult, err := memberRef01Ent.List(memberRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		memberRef01List, memberRef01ListOk := memberRef01ListResult.([]any)
		if !memberRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", memberRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(memberRef01List), map[string]any{"id": memberRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		memberRef01DataUp0Up := map[string]any{
			"id": memberRef01Data["id"],
			"organization_id": setup.idmap["organization_id"],
		}

		memberRef01MarkdefUp0Name := "aaEmail"
		memberRef01MarkdefUp0Value := fmt.Sprintf("Mark01-member_ref01_%d", setup.now)
		memberRef01DataUp0Up[memberRef01MarkdefUp0Name] = memberRef01MarkdefUp0Value

		memberRef01ResdataUp0Result, err := memberRef01Ent.Update(memberRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		memberRef01ResdataUp0 := core.ToMapAny(entityData(memberRef01ResdataUp0Result))
		if memberRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if memberRef01ResdataUp0["id"] != memberRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if memberRef01ResdataUp0[memberRef01MarkdefUp0Name] != memberRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", memberRef01MarkdefUp0Name, memberRef01ResdataUp0[memberRef01MarkdefUp0Name])
		}

		// LOAD
		memberRef01MatchDt0 := map[string]any{
			"id": memberRef01Data["id"],
		}
		memberRef01DataDt0Loaded, err := memberRef01Ent.Load(memberRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		memberRef01DataDt0LoadResult := core.ToMapAny(entityData(memberRef01DataDt0Loaded))
		if memberRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if memberRef01DataDt0LoadResult["id"] != memberRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		memberRef01MatchRm0 := map[string]any{
			"id": memberRef01Data["id"],
		}
		_, err = memberRef01Ent.Remove(memberRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		memberRef01MatchRt0 := map[string]any{
			"organization_id": setup.idmap["organization01"],
		}

		memberRef01ListRt0Result, err := memberRef01Ent.List(memberRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		memberRef01ListRt0, memberRef01ListRt0Ok := memberRef01ListRt0Result.([]any)
		if !memberRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", memberRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(memberRef01ListRt0), map[string]any{"id": memberRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func memberBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "member", "MemberTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read member test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse member test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"member01", "member02", "member03", "action01", "action02", "action03", "board01", "board02", "board03", "card01", "card02", "card03", "enterpris01", "enterpris02", "enterpris03", "notification01", "notification02", "notification03", "organization01", "organization02", "organization03", "token01", "token02", "token03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TRELLO_TEST_MEMBER_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TRELLO_TEST_MEMBER_ENTID": idmap,
		"TRELLO_TEST_LIVE":      "FALSE",
		"TRELLO_TEST_EXPLAIN":   "FALSE",
		"TRELLO_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["TRELLO_TEST_MEMBER_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add organization_id alias for update test.
	if idmapResolved["organization_id"] == nil {
		idmapResolved["organization_id"] = idmapResolved["organization01"]
	}

	if env["TRELLO_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["TRELLO_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewTrelloSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TRELLO_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TRELLO_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: js/test/entity/member/MemberEntity.test.js

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

const Path = require('node:path')
const Fs = require('node:fs')

const { test, describe, afterEach } = require('node:test')
const assert = require('node:assert')


const { TrelloSDK, BaseFeature, stdutil, config } = require('../../..')

const {
  envOverride,
  liveClientOptions,
  liveDelay,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
} = require('../../utility')


describe('MemberEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when TRELLO_TEST_LIVE=TRUE.
  afterEach(liveDelay('TRELLO_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = TrelloSDK.test()
    const ent = testsdk.Member()
    assert(null != ent)
  })


  test('basic', async () => {

    const setup = basicSetup()
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select


    // CREATE
    const member_ref01_ent = client.Member()
    let member_ref01_data = setup.data.new.member['member_ref01']
    member_ref01_data['action_id'] = setup.idmap['action01']
    member_ref01_data['board_id'] = setup.idmap['board01']
    member_ref01_data['enterpris_id'] = setup.idmap['enterpris01']
    member_ref01_data['organization_id'] = setup.idmap['organization01']
    member_ref01_data['token_id'] = setup.idmap['token01']

    member_ref01_data = (await member_ref01_ent.create(member_ref01_data)).data()
    assert(null != member_ref01_data.id)


    // LIST
    const member_ref01_match = {}
    member_ref01_match['organization_id'] = setup.idmap['organization01']

    const member_ref01_list = (await member_ref01_ent.list(member_ref01_match)).map((e) => e.data())

    assert(!isempty(select(member_ref01_list, { id: member_ref01_data.id })))


    // UPDATE
    const member_ref01_data_up0 = {}
    member_ref01_data_up0.id = member_ref01_data.id
    member_ref01_data_up0 ['organization_id'] = setup.idmap['organization_id']

    const member_ref01_markdef_up0 = { name: 'aaEmail', value: 'Mark01-member_ref01_' + setup.now }
    member_ref01_data_up0 [member_ref01_markdef_up0.name] = member_ref01_markdef_up0.value

    const member_ref01_resdata_up0 = (await member_ref01_ent.update(member_ref01_data_up0)).data()
    assert(member_ref01_resdata_up0.id === member_ref01_data_up0.id)

    assert(member_ref01_resdata_up0[member_ref01_markdef_up0.name] === member_ref01_markdef_up0.value)


    // LOAD
    const member_ref01_match_dt0 = {}
    member_ref01_match_dt0.id = member_ref01_data.id
    const member_ref01_data_dt0 = (await member_ref01_ent.load(member_ref01_match_dt0)).data()
    assert(member_ref01_data_dt0.id === member_ref01_data.id)


    // REMOVE
    const member_ref01_match_rm0 = {}
    member_ref01_match_rm0.id = member_ref01_data.id
    await member_ref01_ent.remove(member_ref01_match_rm0)
  

    // LIST
    const member_ref01_match_rt0 = {}
    member_ref01_match_rt0['organization_id'] = setup.idmap['organization01']

    const member_ref01_list_rt0 = (await member_ref01_ent.list(member_ref01_match_rt0)).map((e) => e.data())

    assert(isempty(select(member_ref01_list_rt0, { id: member_ref01_data.id })))


  })
})



function basicSetup(extra) {
  // TODO: fix test def options
  const options = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname,
      '../../../../.sdk/test/entity/member/MemberTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = TrelloSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['member01','member02','member03','action01','action02','action03','board01','board02','board03','card01','card02','card03','enterpris01','enterpris02','enterpris03','notification01','notification02','notification03','organization01','organization02','organization03','token01','token02','token03','enterpris01','enterpris02','enterpris03','member01','member02','member03','organization01','organization02','organization03','member01','member02','member03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'TRELLO_TEST_MEMBER_ENTID': idmap,
    'TRELLO_TEST_LIVE': 'FALSE',
    'TRELLO_TEST_EXPLAIN': 'FALSE',
    'TRELLO_APIKEY': '',
  })

  idmap = env['TRELLO_TEST_MEMBER_ENTID']

  if ('TRUE' === env.TRELLO_TEST_LIVE) {
    client = new TrelloSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
        apikey: env.TRELLO_APIKEY,
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when
      // the last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey and
      // server values above and handed the SDK undefined.
      extra || {}
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.TRELLO_TEST_EXPLAIN,
    now: Date.now(),
  }

  return setup
}
  
